illumos-utils/src/zfs.rs

```diff
@@ -236,56 +237,123 @@ pub struct DatasetProperties {
+        let name_prop_val_source_list = stdout.trim().split('\n');
 
-    let id = match iter.next().context("Missing UUID")? {
-        "-" => None,
-        anything_else => Some(anything_else.parse::<DatasetUuid>()?),
-    };
+        let mut datasets: BTreeMap<&str, BTreeMap<&str, _>> = BTreeMap::new();
+        for name_prop_val_source in name_prop_val_source_list {
+            let mut iter = name_prop_val_source.split_whitespace();
 
-    let name = iter.next().context("Missing 'name'")?.to_string();
-    let avail =
-        iter.next().context("Missing 'avail'")?.parse::<u64>()?.try_into()?;
-    let used =
-        iter.next().context("Missing 'used'")?.parse::<u64>()?.try_into()?;
-    let quota = match iter.next().context("Missing 'quota'")?.parse::<u64>()? {
-        0 => None,
-        q => Some(q.try_into()?),
-    };
-    let reservation =
-        match iter.next().context("Missing 'reservation'")?.parse::<u64>()? {
-            0 => None,
-            r => Some(r.try_into()?),
-        };
-    let compression = iter.next().context("Missing 'compression'")?.to_string();
-
-    Ok(DatasetProperties {
-        id,
-        name,
-        avail,
-        used,
-        quota,
-        reservation,
-        compression,
-    })
-}
+            let (name, prop, val, source) = (
+                iter.next().context("Missing 'name'")?,
+                iter.next().context("Missing 'property'")?,
+                iter.next().context("Missing 'value'")?,
+                iter.next().context("Missing 'source'")?,
+            );
 
-impl FromStr for DatasetProperties {
-    type Err = anyhow::Error;
+            let props = datasets.entry(name).or_default();
+            props.insert(prop, (val, source));
+        }
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        dataset_properties_parse(s)
-            .with_context(|| format!("Failed to parse: {s}"))
+        datasets
+            .into_iter()
+            .map(|(dataset_name, props)| {
+                // Dataset UUIDs are properties that are optionally attached to
+                // datasets. However, some datasets are nested - to avoid them
+                // from propagating, we explicitly ignore this value if it is
+                // inherited.
+                //
+                // This can be the case for the "zone" filesystem root, which
+                // can propagate this property to a child zone without it set.
+                let id = match props.get("oxide:uuid") {
+                    Some((prop, source)) => {
+                        if *source == "inherited" {
+                            None
+                        } else {
+                            Some(
+                                prop.parse::<DatasetUuid>()
+                                    .context("Failed to parse UUID")?,
+                            )
+                        }
+                    }
+                    None => None,
+                };
```

Comment:
Slightly shorter / more consistent with the other properties below, but maybe slightly too clever (abusing the fact that `Option` is an iterator so we can call `.filter()`), so feel free to take it or leave it:

```rust
let id = props
    .get("oxide:uuid")
    .filter(|(_, source)| !source.starts_with("inherited"))
    .map(|(prop, _)| prop.parse::<DatasetUuid>().context("Failed to parse UUID"))
    .transpose()?;
```

Could use the same style for the other properties that have `"default"` values that should be ignored below, too. I wonder if it'd be worth it to wrap `props` in some type that has a `get_unless("property", |source| ...)` helper method for these properties?

Reply:
Sure thing, happy to update this to use `filter`, here and below. I don't like mixing `map` and `match` either, but finding the right magical methods to pass around errors with `map` is sometimes a struggle